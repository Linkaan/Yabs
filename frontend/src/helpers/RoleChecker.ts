import UsersModule, { User } from '../store/modules/UsersModule';

export default class RoleChecker {

  private static checkPermission(user: User, required: number): boolean {
    return user && ((user.role & required) == required || (user.role > required));
  }
  
  static isPrincipal(): boolean {
    return this.checkPermission(UsersModule.currentUser, 32);
  }

  static isAdmin(): boolean {
    return this.checkPermission(UsersModule.currentUser, 16);
  }

  static isCaretaker(): boolean {
    return this.checkPermission(UsersModule.currentUser, 8);
  }

  static isTeacher(): boolean {
    return this.checkPermission(UsersModule.currentUser, 4);
  }

  static isStudentHealth(): boolean {
    return this.checkPermission(UsersModule.currentUser, 2);
  }

  static isStudent(): boolean {
    return UsersModule.currentUser && (UsersModule.currentUser.role & 1) == 1;
  }
}
