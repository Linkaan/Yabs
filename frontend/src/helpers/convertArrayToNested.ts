export default function convertList(payload: any[], typeOfId: string): object { //eslint-disable-line
  return payload.reduce((acc, item) => {
    acc[item[typeOfId]] = item;
    return acc;
  }, {});
}
