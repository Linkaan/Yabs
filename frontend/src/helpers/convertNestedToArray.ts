export default function convertList(payload: any, typeOfId: string): any[] { //eslint-disable-line
  const res = Object.keys(payload).map((id) => payload[id]);
  return res;
}
