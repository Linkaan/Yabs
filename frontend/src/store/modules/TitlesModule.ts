import Vue from 'vue';
import {
  VuexModule,
  Module,
  getModule,
  Action,
  Mutation,
} from 'vuex-module-decorators';
import store from '..';
import TitlesAPI from '../../api/titles';
import convertList from '../../helpers/convertArrayToNested';

export interface Title {
  cost: number;
  created_at: string;
  id: number;
  isbn: string;
  name: string;
  title_type: string;
  updated_at: string;
}

export interface TitleForm {
  name: string;
  cost: string;
  isbn: string;
  title_type: string;
}

export interface TitleCollection { [id: number]: Title; }

@Module({dynamic: true, namespaced: true, name: 'TitlesModule', store})
class TitlesModule extends VuexModule {
  private _failure: any = null;
  private _titles: TitleCollection = {};

  get all() {
    return this._titles;
  }

  get allAsArray() {
    return Object.keys(this._titles).map( (id) => this._titles[Number(id)]);
  }

  @Action({rawError: true})
  public fetchAll() {
    return new Promise((resolve, reject) => {
      TitlesAPI.all()
        .then((response: any) => {
          this.convertTitleList(response);
          resolve();
        })
        .catch((error: any) => {
          this.setFailure(error);
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public create(request: TitleForm) {
    return new Promise((resolve, reject) => {
      TitlesAPI.create(request)
        .then((response: any) => {
          this.setTitle(response);
          resolve(response);
        })
        .catch((error: any) => {
          this.setFailure(error);
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public update(request: any) {
    return new Promise((resolve, reject) => {
      TitlesAPI.update(request)
        .then((response: any) => {
          this.setTitle(response);
          resolve(response);
        })
        .catch((error: any) => {
          this.setFailure(error);
          reject(error);
        });
    });
  }

  @Action({rawError: true})
  public delete(request: any) {
    return new Promise((resolve, reject) => {
      TitlesAPI.delete(request)
        .then((response: any) => {
          this.removeTitle(response);
          resolve(response);
        })
        .catch((error: any) => {
          this.setFailure(error);
          reject(error);
        });
    });
  }

  @Mutation
  private setTitle(payload: any) {
    Vue.set(this._titles, payload.id, payload);
  }

  @Mutation
  private removeTitle(titleId: string) {
    Vue.delete(this._titles, titleId);
  }

  @Mutation
  private setFailure(payload: any) {
    this._failure = payload;
  }

  @Mutation
  private convertTitleList(payload: Title[]) {
    const list = convertList(payload, 'id');
    this._titles = list;
  }
}

export default getModule(TitlesModule);
