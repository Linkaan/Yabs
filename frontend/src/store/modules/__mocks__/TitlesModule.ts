import { Title } from '@/types';

const title: Title = {
  'id': 1,
  'name': 'Alice in Wonderland',
  'isbn': '0763645680',
  'cost': 700,
  'title_type': 'SkÃ¶nlitteratur',
  'created_at': '2020-12-21T15:23:15.615Z',
  'updated_at': '2021-01-04T15:54:04.695Z',
};

const title1: Title = {
  'id': 3,
  'name': 'Clean Code',
  'isbn': '9780132350884',
  'cost': 300,
  'title_type': 'Kurslitteratur',
  'created_at': '2020-12-21T15:23:15.652Z',
  'updated_at': '2020-12-21T15:23:15.652Z',
};

module.exports = {
  allAsArray: [title, title1],
  fetchAll: (): null => null,
  create: jest.fn().mockImplementationOnce(_arg => {
    return Promise.resolve();
  }),
  update: jest.fn(),
  fetchSingle: (_: any): Title => { // eslint-disable-line @typescript-eslint/no-explicit-any
    return title;
  },
  fetchSingleByISBN: jest.fn().mockImplementation(() => {
    return new Promise((resolve, _) => {
      resolve(title);
    });
  }),
};
