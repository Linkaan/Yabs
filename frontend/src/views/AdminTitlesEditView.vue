<template>
  <div>
    <h1 class="display-2">
      Titles
    </h1>
    <div>
      <TitleFormComponent
        :form-data="data"
        mode="edit"
      />
    </div>
  </div>
</template>

<script lang="ts">
import { ref, defineComponent, SetupContext } from '@vue/composition-api';
import TitlesModule from '../store/modules/TitlesModule';
import TitleFormComponent from '@/components/TitleFormComponent.vue';
import { Title } from '../types';

export default defineComponent({
  name: 'AdminTitlesEditView',
  components: {
    TitleFormComponent
  },
  setup(_: object, { root }: SetupContext) {
    const data = ref();
    TitlesModule.fetchSingleByISBN(root.$route.params.id).then((response: Title) => {
      data.value = response;
    });
    return { data };
  }
});
</script>

<style scoped>
</style>
