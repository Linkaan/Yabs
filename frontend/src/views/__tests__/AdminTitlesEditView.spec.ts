import AdminTitlesEditView from '../AdminTitlesEditView.vue';
import vue from 'vue';

import { VueClass, Wrapper } from '@vue/test-utils';
import { factory, shallowFactory } from '@/helpers/testFactoryHelpers';
import TitlesModule from '@/store/modules/TitlesModule';

jest.mock('@/store/modules/TitlesModule');

const options = {
  mocks: {
    $route: {
      params: {
        id: '0763645680'
      }
    }
  },
};

describe('AdminTitlesEditView.vue', () => {
  it('Displays the correct data in the form', async () => {
    const wrapper= shallowFactory(AdminTitlesEditView as VueClass<Vue>, options);
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.text()).not.toMatch('Clean Code');
  });

  it('Calls titles module with correct data', async () => {
    const wrapper= factory(AdminTitlesEditView as VueClass<Vue>, options);
    wrapper.find('[data-jest=deleteTitle]').trigger('click');
    await expect(TitlesModule.fetchSingleByISBN).toHaveBeenCalledWith('0763645680');
    expect(TitlesModule.getLoans).toHaveBeenCalledWith(1);
  });
});
