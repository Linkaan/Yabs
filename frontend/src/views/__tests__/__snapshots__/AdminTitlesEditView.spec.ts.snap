// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminTitlesEditView.vue Displays the correct data in the form 1`] = `
<div>
  <h1 class="display-2">
    Titles
  </h1>
  <div>
    <form novalidate="novalidate" data-jest="form" class="v-form">
      <div class="v-item-group theme--light" vertical="" label-class="text-sm-right" label-for="nestedBid">
        <div class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
          <div class="v-input__control">
            <div class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span>​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><label for="nestedBid" class="v-label theme--light" style="left: 0px; position: absolute;">Namn</label><input data-jest="name" data-cy="name" id="nestedBid" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-item-group theme--light" vertical="" label-class="text-sm-right" label-for="nestedBid">
        <div class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
          <div class="v-input__control">
            <div class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span>​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><label for="nestedBid" class="v-label theme--light" style="left: 0px; position: absolute;">Kostnad</label><input data-jest="cost" data-cy="cost" id="nestedBid" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-item-group theme--light" vertical="" label-class="text-sm-right" label-for="nestedBid">
        <div class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
          <div class="v-input__control">
            <div class="v-input__slot">
              <fieldset aria-hidden="true">
                <legend style="width: 0px;"><span>​</span></legend>
              </fieldset>
              <div class="v-text-field__slot"><label for="nestedBid" class="v-label theme--light" style="left: 0px; position: absolute;">isbn</label><input data-jest="isbn" data-cy="isbn" id="nestedBid" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-21" class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span>​</span></legend>
            </fieldset>
            <div class="v-select__slot"><label for="input-21" class="v-label theme--light" style="left: 0px; position: absolute;">Välj typ av titel</label>
              <div class="v-select__selections"><input data-jest="titleType" data-cy="titleType" id="input-21" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
          </div>
        </div>
      </div>
      <div class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
        <div class="v-input__control">
          <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-26" class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend style="width: 0px;"><span>​</span></legend>
            </fieldset>
            <div class="v-select__slot"><label for="input-26" class="v-label theme--light" style="left: 0px; position: absolute;">Välj ämne av titel</label>
              <div class="v-select__selections"><input data-jest="subjectId" data-cy="subjectId" id="input-26" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
              </div><input type="hidden">
            </div>
            <div class="v-menu">
              <!---->
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
          </div>
        </div>
      </div> <button type="submit" class="mr-4 v-btn v-btn--flat theme--light v-size--large primary--text" data-cy="submit"><span class="v-btn__content">
    Uppdatera titel
  </span></button> <button type="reset" class="v-btn v-btn--flat theme--light v-size--large" data-cy="reset"><span class="v-btn__content">
    Rensa Fältet
  </span></button>
    </form>
  </div>
</div>
`;
